File: Python/Hash Maps/HashTable.py
class HashTable:
    def __init__(self, size):
        self.keyMap = [None] * size

    def _hash(self, key):
        total = 0
        WEIRD_PRIME = 53
        for i in range(0, min(len(key), 100)):
            char = key[i]
            value = ord(char) - 96
            total = (total * WEIRD_PRIME + value) % len(self.keyMap)
        return total

    def set(self, key, value):
        hash = self._hash(key)
        if self.keyMap[hash] == None:
            self.keyMap[hash] = []
        self.keyMap[hash].append([key, value])
        return self.keyMap


    def get(self, key):
        hash = self._hash(key)
        if self.keyMap[hash] == None:
            return None
        else:
            for i in range(0, len(self.keyMap[hash])):
                if self.keyMap[hash][i][0] == key:
                    return self.keyMap[hash][i][1]
        return None

    def keys(self):
        keysArr = []
        for i in range(0, len(self.keyMap)):
            if self.keyMap[i] != None:
                for j in range(0, len(self.keyMap[i])):
                    keysArr.append(self.keyMap[i][j][0])
        return keysArr

    def values(self):
        valuesArr = []
        for i in range(0, len(self.keyMap)):
            if self.keyMap[i] != None:
                for j in range(0, len(self.keyMap[i])):
                    valuesArr.append(self.keyMap[i][j][1])
        return valuesArr

map = HashTable(13)
print(map.set("pink", "#ff3092"))
print(map.set("blue", "#a03e5c"))
print(map.set("green", "#bb3092"))
print(map.set("cyan", "#ae3e5f"))
print(map.set("black", "#6fc092"))
print(map.set("white", "#a69c5v"))
print(map.set("orange", "#ffe354"))
print(map.set("yellow", "#00ff5c"))
print(map.set("maroon", "#00ff5c"))
print(map.get("green"))
print(map.get("pink"))  
print(map.get("orange"))
print(map.get("cyan"))
print(map.get("blue"))
print(map.get("yellow"))
print(map.get("sky"))
print(map.keys())
print(map.values())

File: JavaScript/Hash Maps/HashTable.js
class HashTable {
    constructor(size) {
        this.keyMap = new Array(size);
    }

    _hash(key) {
        let total = 0;
        let WEIRD_PRIME = 53;
        for (let i = 0; i < Math.min(key.length, 100); i++) {
            let char = key[i];
            let value = char.charCodeAt(0) - 96;
            total = (total * WEIRD_PRIME + value) % this.keyMap.length;
        }
        return total;
    }

    set(key, value) {
        let hash = this._hash(key);
        if (!this.keyMap[hash]) {
            this.keyMap[hash] = [];
        }
        this.keyMap[hash].push([key, value]);
        return this.keyMap;
    }

    get(key) {
        let hash = this._hash(key);
        if (this.keyMap[hash]) {
            for (let i = 0; i < this.keyMap[hash].length; i++) {
                if (this.keyMap[hash][i][0] === key) {
                    return this.keyMap[hash][i][1];
                }
            }
        }
        return;
    }

    keys() {
        let keysArr = [];
        for (let i = 0; i < this.keyMap.length; i++) {
            if (this.keyMap[i]) {
                for (let j = 0; j < this.keyMap[i].length; j++) {
                    if (!keysArr.push(this.keyMap[i][j][0])) {
                        keysArr.push(this.keyMap[i][j][0]);
                    }
                }
            }
        }
        return keysArr;
    }

    values() {
        let valuesArr = [];
        for (let i = 0; i < this.keyMap.length; i++) {
            if (this.keyMap[i]) {
                for (let j = 0; j < this.keyMap[i].length; j++) {
                    if (!valuesArr.includes(this.keyMap[i][j][1])) {
                        valuesArr.push(this.keyMap[i][j][1]);
                    }
                }
            }
        }
        return valuesArr;
    }
}

let map = new HashTable(13);
console.log(map.set("pink", "#ff3092"));
console.log(map.set("blue", "#a03e5c"));
console.log(map.set("green", "#bb3092"));
console.log(map.set("cyan", "#ae3e5f"));
console.log(map.set("black", "#6fc092"));
console.log(map.set("white", "#a69c5v"));
console.log(map.set("orange", "#ffe354"));
console.log(map.set("yellow", "#00ff5c"));
console.log(map.set("maroon", "#00ff5c"));
console.log(map.get("green"));
console.log(map.get("pink"));
console.log(map.get("orange"));
console.log(map.get("cyan"));
console.log(map.get("blue"));
console.log(map.get("yellow"));
console.log(map.get("sky"));
console.log(map.keys());
console.log(map.values());